tech-spec.md

```diff
@@ -0,0 +1,310 @@
+    ```python
+    extract(data_source: str, schema_file: str, org: str=''
+            ) -> Dict[TableName, Dataframe]
+    ```
+    - data_source: a file path or database url (in SQLAlchemy format)
+    - schema_file: rule schema file path
+    - org: org to share with, or all orgs if empty
+    - returns a Pandas Dataframe per table, for further processing
+- low level:
+    - connect(data_source: str) -> Connection
+    - parse(schema_file: str) -> Dict[OrgName, Dict[TableName, Query]]
+    - extract(Connection, Query) -> Dataframe
+    - getCounts(Connection, Query) -> Dict[RuleId, int]
+    - getColumns(Connection, Query) -> Tuple[RuleId, List[ColumnName]]
+
+### Private modules
+
+#### cons
+
+Data source connection abstraction, including intermediate copy of spreadsheet
+files to in-memory DBs.
+
+- connect(data_source: str) -> Connection
+
+#### rules
+
+Loading of rule files.
+
+- load(schema_file: str) -> Dict[RuleId, Rule]
+
+#### trees
+
+Parsing of rules into abstract syntax trees.
+
+- parse(rules: Dict[RuleId, Rule] | List[Rule]) -> RuleTree
+
+#### queries
+
+(SQL) query generation from ASTs.
+
+- generate(rt: RuleTree) -> Dict[OrgName, Query]
+
+### Examples
+
+common definitions:
+```python
+import pandas as pd
+
+import sharing as s
+
+data_file = 'data.xlsx'
+rule_file = 'rules.csv'
+org = 'OHRI'
+```
+
+high-level one-shot function:
+
+```python
+results = s.extract(data_file, rules, org)
+for table, data in results.items():
+    data.to_csv(f'{org}-{table}.csv')
+```
+
+low-level sample code:
+
+```python
+def describeTableQuery(con, rules, table, query):
+    print(f'query table: {table}')
+
+    (selectRuleId, columns) = s.getColumns(con, query)
+    print(f'query columns (from rule {selectRuleId}):')
+    print(','.join(columns))
+
+    print('query counts per rule:')
+    ruleCounts = s.getCounts(con, query)
+    for ruleId, count in ruleCounts.items():
+        r = rules[ruleId]
+        ruleFilter = f'{r.key} {r.operator} {r.value}'
+        print(f'{ruleId} | {count} | {ruleFilter}')
+
+def extractFilteredData(con, table, query):
+    data: pd.Dataframe = s.extract(con, query)
+    data.to_csv(f'{org}-{table}.csv')
+
+con = s.connect(data_file)
+rules = s.load(rule_file)
+ruleTree = s.parse(rules)
+tableQueries = s.generate(ruleTree)[org]
+for table, query in tableQueries.items():
+    describeTableQuery(con, rules, table, query)
+    extractFilteredData(con, table, query)
+```
+
+## Rule schema parsing
+
+1. open schema file
+2. read lines
+3. parse each line into a rule obj
+4. add each rule obj to a dictionary with rule id as key
+
+## AST generation
+
+Node kinds:
+- **root**: AST root
+- **share**: what to share with each org
+- **table**: table name
+- **select**: lists column name values
+- **group**: groups filters together
+- **filter**: defines a filter with operator, key and value
+- **key**: field name
+- **value**: generic
+
+Node structure:
+
+- (ruleId: int)
+- kind: NodeKind
+- value: str
```

Comment:
good point. fixed:

- NodeKind.key -> field (since it's a field name)
- NodeKind.value -> literal (since it's a string literal)
- Node.value -> str_val (since it's the string representation of the actual value)